crates/telio-firewall/src/firewall.rs

```diff
@@ -608,15 +644,71 @@ impl StatefullFirewall {
     }
 
     fn handle_inbound_icmp<'a, P: IpPacket<'a>>(&self, peer: &PublicKey, ip: &P) -> bool {
-        let icmp_packet = unwrap_option_or_return!(P::Icmp::new(ip.payload()), false);
+        let icmp_packet = unwrap_option_or_return!(IcmpPacket::new(ip.payload()), false);
 
-        if P::Icmp::BLOCKED_TYPES.contains(&icmp_packet.get_type()) {
+        if P::Icmp::BLOCKED_TYPES.contains(&icmp_packet.get_icmp_type().0) {
             telio_log_trace!("Dropping ICMP packet {:?} {:?}", ip, peer);
             return false;
         }
 
-        telio_log_trace!("Accepting ICMP packet {:?} {:?}", ip, peer);
-        true
+        let mut icmp_cache = unwrap_lock_or_return!(self.icmp.lock(), false);
+        let key = unwrap_option_or_return!(Self::build_icmp_key(ip, true), false);
+        if icmp_cache.get(&key).is_some() {
+            telio_log_trace!("Matched ICMP conntrack entry {:?}", key,);
+            telio_log_trace!("Removing ICMP conntrack entry {:?}", key);
+            icmp_cache.remove(&key);
+            telio_log_trace!("Accepting ICMP packet {:?} {:?}", ip, peer);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn build_icmp_key<'a, P: IpPacket<'a>>(ip: &P, inbound: bool) -> Option<IcmpConn> {
+        let icmp_packet = unwrap_option_or_return!(IcmpPacket::new(ip.payload()), None);
+        let (response_type, has_identifier_sequence) = match icmp_packet.get_icmp_type().0 {
+            // Request messages get the icmp type of the corresponding reply message
+            // Reply messages get their own icmp type, to match that of its corresponding request
+            // ICMP types
+            0 => (0, true),   // echo reply message
+            8 => (0, true),   // echo request message
+            13 => (14, true), // timestamp message
+            14 => (14, true), // timestamp reply message
+            15 => (16, true), // information request message
+            16 => (16, true), // information reply message
+            17 => (18, true), // address mask request message
+            18 => (18, true), // address mask reply message
+            // ICMPv6 types
+            128 => (129, true),  // echo request message
+            129 => (129, true),  // echo reply message
+            133 => (134, false), // router solicit message
+            134 => (134, false), // router advert message
+            135 => (136, false), // neighbor solicit message
+            136 => (136, false), // neighbor advert message
+            // No other types go into the cache
+            _ => return None,
+        };
+        let identifier_sequence = if has_identifier_sequence {
+            let bytes = icmp_packet.payload()[0..4].try_into();
```

Comment:
I think you've again forgot about the check for the length of the payload?

Reply:
`try_into` returns a `Result` with an `Err` if the body is not long enough, which then gets rejected by the if-else below this line

Reply:
It does not protect you from the panic. `try_into` in this example converts from slice to array. It converts to 4 element array because that's what expected by `u32::from_ne_bytes` and it will fail if you give it a slice of the wrong size. E.g. `icmp_packet.payload()[0..2].try_into()` will return None because the slice is too short. But it does nothing to check that your slice itself is correct, and in that case it might not be correct. So if your slice points to unallocated memory this will panic.

Reply:
Here's the playground to verify: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=e8bcc5f5b43d0e32c7f62afb6d8522c6

Reply:
I didn't consider the slicing operation, only the conversion. You are of course correct, and I will add a check